google-cloud-spanner/src/main/java/com/google/cloud/spanner/TransactionContext.java

```diff
@@ -130,16 +131,19 @@
    * 2nd statement, and an array of length 1 that contains the number of rows modified by the 1st
    * statement. The 3rd statement will not run.
    */
-  long[] batchUpdate(Iterable<Statement> statements);
+  long[] batchUpdate(Iterable<Statement> statements, UpdateOption... options);
 
   /**
-   * Same as {@link #batchUpdate(Iterable)}, but is guaranteed to be non-blocking. If multiple
-   * asynchronous update statements are submitted to the same read/write transaction, the statements
-   * are guaranteed to be submitted to Cloud Spanner in the order that they were submitted in the
-   * client. This does however not guarantee that an asynchronous update statement will see the
-   * results of all previously submitted statements, as the execution of the statements can be
-   * parallel. If you rely on the results of a previous statement, you should block until the result
-   * of that statement is known and has been returned to the client.
+   * Same as {@link #batchUpdate(Iterable, UpdateOption...)}, but is guaranteed to be non-blocking.
+   * If multiple asynchronous update statements are submitted to the same read/write transaction,
+   * the statements are guaranteed to be submitted to Cloud Spanner in the order that they were
+   * submitted in the client. This does however not guarantee that an asynchronous update statement
+   * will see the results of all previously submitted statements, as the execution of the statements
+   * can be parallel. If you rely on the results of a previous statement, you should block until the
+   * result of that statement is known and has been returned to the client.
    */
-  ApiFuture<long[]> batchUpdateAsync(Iterable<Statement> statements);
+  ApiFuture<long[]> batchUpdateAsync(Iterable<Statement> statements, UpdateOption... options);
+
+  /** Sets the tag for the currently active transaction. */
+  void withTransactionTag(String tag);
```

Comment:
I've been working on adding support for RPC priority, which will use the same `RequestOptions` proto as this change, and has many similarities with tagging. I'm starting to doubt whether this is the best way to support transaction tags, as this approach would also mean that we would have to add a `withPriority(..)` method for RPC priority, and possibly additional methods for any future additional request options.
It is also a little bit counter-intuitive that the transaction tag is set using `TransactionOptions` for the `writeWithOptions` and `writeAtLeastOnceWithOptions`, while it cannot be used for normal read/write transactions.

I therefore created the change for RPC priority by adding `TransactionOption... options` to the read/write transaction methods instead of adding a `withPriority(...)` method to `TransactionContext`.

See here for the definition: https://github.com/googleapis/java-spanner/blob/046aac8e1d04872cf14fe918be70d2433b5150c1/google-cloud-spanner/src/main/java/com/google/cloud/spanner/DatabaseClient.java#L325

See here for an example how it's used: https://github.com/googleapis/java-spanner/blob/046aac8e1d04872cf14fe918be70d2433b5150c1/google-cloud-spanner/src/test/java/com/google/cloud/spanner/DatabaseClientImplTest.java#L1695

I think we should use the same approach for both tagging and RPC priority.
@mayurkale22 @thiagotnunes @syeduguri WDYT?

Reply:
I like the fact that if we do this we would normalise the API (for `writeWithOptions` and `writeAtLeastOnceWithOptions` along with the read write transactions). I think that the problem that arrives with this is that for all the current varargs APIs, the user could potentially set multiple Tags or multiple Priorities. We would need them to add client side validation or choose the first/last one.

I don't have a strong opinion on this, I am good either way, but regardless of what we choose, we should follow the same approach for Tagging and RPC priorities.

Reply:
I think this is a great idea. I can update my PR based on this. 

Looking at #676, I see `TransactionOption` has been added on `TransactionManager  transactionManager()`, `AsyncRunner runAsync()` and `AsyncTransactionManager transactionManagerAsync()`, Could you please elaborate more on this? Also, do I need to make things changes in my PR?

Reply:
Well, it depends on which PR (this or #676) will be merged first. The one that is merged first should add the `TransactionOption` argument to the transaction methods. The other PR should then be rebased after that change and use the same approach for setting the options. I'll try to find out which one is most likely to be ready for merge first.

Reply:
I will rebase this CL once https://github.com/googleapis/java-spanner/pull/716 is merged.